### PR TITLE
Preserve order when regenerating GPU images

### PR DESCRIPTION
**Pull request by dnfield**
Fixes https://github.com/flutter/flutter/issues/122566

The cause of that bug is that multiple dependent `toImageSync` calls are made such that each successive image depends on the previous one.

In the current implementation, a `std::map` is causing the `OnGrContextCreated` calls to happen in sort-order based on pointer values. This patch adds another map to track the insertion order and a test to make sure that the callbacks respect insertion order.

### CODE REVIEW

File: common/graphics/texture.cc [hunk 1]
```diff
@@ -26,7 +26,13 @@ void TextureRegistry::RegisterTexture(const std::shared_ptr<Texture>& texture) {
 void TextureRegistry::RegisterContextListener(
     uintptr_t id,
     std::weak_ptr<ContextListener> image) {
-  images_[id] = std::move(image);
+  size_t next_id = image_counter_.fetch_add(1);
```
Comment (chinmaygarde) [inline common/graphics/texture.cc:29]: C++ operator overloading of `++` does this for you already. IMO that is more readable that `fetch_add`.

File: common/graphics/texture.cc [hunk 2]
```diff
@@ -61,11 +77,14 @@ void TextureRegistry::OnGrContextDestroyed() {
...
-  for (const auto& [id, weak_image] : images_) {
+  for (const auto& [id, pair] : ordered_images_) {
+    auto index_id = pair.first;
+    auto weak_image = pair.second;
     if (auto image = weak_image.lock()) {
       image->OnGrContextDestroyed();
     } else {
-      images_.erase(id);
+      image_indices_.erase(index_id);
+      ordered_images_.erase(id);
```
Comment (chinmaygarde) [inline common/graphics/texture.cc:87]: Here and elsewhere, mutation while enumeration makes me nervous. Are iterators on `ordered_images_` invalidated when the erase happens here?

Comment (chinmaygarde) [reply common/graphics/texture.cc:87]: Used to doing this safely by using iterators explicitly and assigning the iterator to the return call of `erase`. I am not sure I have ever done it this way. So want to double check this is safe.

File: common/graphics/texture.cc (hunk 1 above)
Comment (chinmaygarde) [reply common/graphics/texture.cc:29]: Besides, does this need to be an atomic? I thought all texture registry operations were [raster thread only](https://github.com/flutter/engine/blob/6069c8f885a1bd829c26a66df8807165c7ad691c/common/graphics/texture.h#L77).

File: common/graphics/texture.h [hunk 3]
```diff
@@ -95,7 +95,15 @@ class TextureRegistry {
...
   std::map<int64_t, std::shared_ptr<Texture>> mapping_;
-  std::map<uintptr_t, std::weak_ptr<ContextListener>> images_;
+  std::atomic_size_t image_counter_;
+  // This map keeps track of registered context listeners by their own
+  // externally provided id. It indexes into ordered_images_.
+  std::map<uintptr_t, size_t> image_indices_;
+  // This map makes sure that iteration of images happens in insertion order
+  // (managed by image_counter_) so that images which depend on other images get
+  // re-created in the right order.
+  std::map<size_t, std::pair<uintptr_t, std::weak_ptr<ContextListener>>>
```
Comment (chinmaygarde) [inline common/graphics/texture.h:105]: I guess this makes sense but is a bit rough to read. How about we take advantage of the fact that the map is ordered already and store the existing key and the order as a key type. Then you can have a custom std::less on the key type and typedef the whole thing for more readibility. Iterating over the map elements will be order then. I suppose you will need a separate map for the removal (a unintptr to order association).

Just a suggestion.

File: common/graphics/texture.cc (hunk 1 above)
Comment (dnfield) [reply common/graphics/texture.cc:29]: I guess it doesn't need to be atomic. I was being overly cautious :)

File: common/graphics/texture.cc (hunk 2 above)
Comment (dnfield) [reply common/graphics/texture.cc:87]: Yeah, I'll change this. It seems to work but it'd be safer to use the iterator.

I had another issue with mutation during iteration that I fixed by copying the map into a vector, but that would be overkill here.

File: common/graphics/texture.h (hunk 3 above)
Comment (dnfield) [reply common/graphics/texture.h:105]: I considered that but it comes down to either storing extra data in the key or extra data in the value.

This makes the key easier to hash and avoids the need for a custom comparer that ignores part of the data. I'll use the typedef though.

Comment (dnfield) [reply common/graphics/texture.h]: (if I didn't need to worry about updating the order when someone re-registers I could just use a std::vector)

**Review by chinmaygarde — APPROVED** [review]

**Pull request merged**